ml-kem/src/lib.rs

```diff
@@ -0,0 +1,204 @@
+mod crypto;
+
+/// Section 4.2.1. Conversion and Compression Algorithms, Compression and decompression
+mod compress;
+
+/// Section 4.2.1. Conversion and Compression Algorithms, Encoding and decoding
+mod encode;
+
+/// Section 5. The K-PKE Component Scheme
+mod pke;
+
+/// Section 6. The ML-KEM Key-Encapsulation Mechanism
+pub mod kem;
+
+/// Section 7. Parameter Sets
+mod param;
+
+use core::fmt::Debug;
+use crypto_common::rand_core::CryptoRngCore;
+use generic_array::{ArrayLength, GenericArray};
+use typenum::{U10, U11, U2, U3, U4, U5};
+
+#[cfg(feature = "deterministic")]
+pub use util::B32;
+
+pub use param::ParameterSet;
+
+/// An object that knows what size it is
+pub trait EncodedSizeUser {
+    /// The size of an encoded object
+    type EncodedSize: ArrayLength;
+
+    /// Parse an object from its encoded form
+    fn from_bytes(enc: &Encoded<Self>) -> Self;
+
+    /// Serialize an object to its encoded form
+    fn as_bytes(&self) -> Encoded<Self>;
+}
+
+/// A byte array encoding a value the indicated size
+pub type Encoded<T> = GenericArray<u8, <T as EncodedSizeUser>::EncodedSize>;
+
+/// A key that can be used to decapsulated an encapsulated shared key
+pub trait DecapsulationKey<SharedKey, Ciphertext>: EncodedSizeUser {
+    /// Decapsulate the ciphertext to obtain the encapsulated shared key
+    fn decapsulate(&self, ciphertext: &Ciphertext) -> SharedKey;
+}
+
+/// A key that can be used to encapsulate a shared key such that it can only be decapsulated
+/// by the holder of the corresponding decapsulation key
+pub trait EncapsulationKey<SharedKey, Ciphertext>: EncodedSizeUser {
+    /// Encapsulate a fresh secret to the holder of the corresponding decapsulation key
+    fn encapsulate(&self, rng: &mut impl CryptoRngCore) -> (SharedKey, Ciphertext);
+
+    /// Encapsulate a specific shared key to the holder of the corresponding decapsulation key
+    #[cfg(feature = "deterministic")]
+    fn encapsulate_deterministic(&self, m: &B32) -> (SharedKey, Ciphertext);
+}
+
+/// A generic interface to a Key Encapsulation Method
+pub trait KemCore {
+    /// The size of a shared key generated by this KEM
+    type SharedKeySize: ArrayLength;
+
+    /// The size of a ciphertext encapsulating a shared key
+    type CiphertextSize: ArrayLength;
+
+    /// A decapsulation key for this KEM
+    type DecapsulationKey: DecapsulationKey<SharedKey<Self>, Ciphertext<Self>> + Debug + PartialEq;
+
+    /// An encapsulation key for this KEM
+    type EncapsulationKey: EncapsulationKey<SharedKey<Self>, Ciphertext<Self>> + Debug + PartialEq;
+
+    /// Generate a new (decapsulation, encapsulation) key pair
+    fn generate(rng: &mut impl CryptoRngCore) -> (Self::DecapsulationKey, Self::EncapsulationKey);
+
+    /// Generate a new (decapsulation, encapsulation) key pair deterministically
+    #[cfg(feature = "deterministic")]
+    fn generate_deterministic(d: &B32, z: &B32)
+        -> (Self::DecapsulationKey, Self::EncapsulationKey);
+}
```

Comment:
We do have some existing traits for KEMs here:

- https://github.com/RustCrypto/traits/tree/master/kem
- https://docs.rs/kem/latest/kem/

Reply:
Yup, though its tires haven't been thoroughly kicked so I'm curious if @bifurcation thinks it's an OK fit.

Also it probably needs to be upgraded to hybrid-array

Reply:
Oh nice re KEM traits, somehow I had missed that.  Obviously should use that, was already thinking that that trait should be outside this crate.

@tarcieri if you concur re `hybrid_array`, happy to switch to that.  (Assuming there are no roadblocks, e.g., due to some of the typenum craziness we're going here.)  Might do some benchmarking to see if there are perf impacts.

Reply:
To be clear: I haven't taken a deep look at the KEM traits yet, will do and report back.  But it's correct to have the KEM traits outside this crate.

Reply:
We're migrating to `hybrid-array` elsewhere. It should be a drop-in replacement for `generic-array`. There's a migration guide here (it should mostly be find/replace): https://docs.rs/hybrid-array/0.2.0-rc.5/hybrid_array/#migrating-from-generic-array

I can also look at updating the `kem` crate to use `hybrid-array`.

Reply:
A challenge with using `hybrid-array` here is that it lacks functional traits like `Concat` and `Split`.  It should be possible to reimplement those in the obvious way without a ton of trouble, just needs to be done.  @tarcieri I assume you'd be open to a PR on `hybrid-array` to add some stuff like that?

Reply:
There are inherent methods: [`Array::concat`](https://docs.rs/hybrid-array/0.2.0-rc.5/hybrid_array/struct.Array.html#method.concat) and [`Array::split`](https://docs.rs/hybrid-array/0.2.0-rc.5/hybrid_array/struct.Array.html#method.split). Also note that `concat` is easily built on `FromIter`: https://docs.rs/hybrid-array/0.2.0-rc.5/src/hybrid_array/lib.rs.html#239

Reply:
Oh, never mind, I just found those!  Excellent, back to implementing...

Reply:
OK, it looks like the only roadblock into is that `ArraySize` isn't implemented for arbitrary `UInt` values.  In particular, the compiler needs to know that  the following values as `UInt`s implement `ArraySize`:

```
 640 = 2 * (32 * 10)
 960 = 3 * (32 * 10)
1408 = 4 * (32 * 11)
```

Don't know if you want to add specific `ArraySize` implementations to `hybrid_array` for these values, or whether it's possible to make a blanket implementation as `generic_array` does.  Either way, happy to send a PR.

Reply:
The inability to implement traits for arbitrary sizes is definitely a tradeoff of `generic-array` vs `hybrid-array` (see also https://github.com/RustCrypto/hybrid-array/issues/41).

However, recent work on the crate has massively simplified the amount of code required to support a size down to just two very small trait impls, so I can probably increase the base set of size impls up to 4096, and we can potentially add a crate feature to add larger additional sizes.

Reply:
On trying to build tests, I find that I need `ArraySize` implementations for a bunch more sizes: 480, 800, 1088, 1152, 1184, 1536, 1568, 1600, 1632, 2336, 2368, 2400, 3104, 3136, 3168.  So increasing the base set size up to 4096 would cover me.

Given this dependency, we might want to consider migrating to `hybrid-array` as a follow-on to this PR.

Reply:
Yeah, that gets a bit tricky and I see `Const<N>` doesn't cover that range either. It will require encoding the bit pattern for those sizes, and in such a case I could probably consider e.g. multiples of 32 in the range 1024-4096 which should cover your needs at least.

Reply:
> Given this dependency, we might want to consider migrating to hybrid-array as a follow-on to this PR.

Yes, agreed. Sorry for the unnecessary detour. I just started trying to migrate the crate myself and hit the same large-const issues for Saber.

Reply:
No worries @rozbb!  It does look a bit nicer than `generic-array`, so I like it as a long-term direction.  It's unfortunate that the state of const generics is such that it requires all this hacking around.

Reply:
This should at least be a stopgap to support the sizes you need with `hybrid-array`: https://github.com/RustCrypto/hybrid-array/pull/49/files

I'll see if I can come up with a better/consistent/more automated way of implementing these additional sizes.